pkg/controller/persistentvolume/persistentvolume_claim_binder_controller.go

```diff
@@ -282,20 +283,34 @@ func syncClaim(volumeIndex *persistentVolumeOrderedIndex, binderClient binderCli
 			return fmt.Errorf("A volume match does not exist for persistent claim: %s", claim.Name)
 		}
 
-		// make a binding reference to the claim.
-		// triggers update of the claim in this controller, which builds claim status
+		// create a reference to the claim and assign it to the volume being bound.
+		// the volume is a pointer and assigning the reference fixes a race condition where another
+		// claim might match this volume but before the claimRef is persistent in the next case statement
+		claimRef, err := api.GetReference(claim)
+		if err != nil {
+			return fmt.Errorf("Unexpected error getting claim reference: %v\n", err)
+		}
+
+		// make a binding reference to the claim and ensure to update the local index to prevent dupe bindings
+		clone, err := conversion.NewCloner().DeepCopy(volume)
```

Comment:
You use this in three places.  You mind making something like https://github.com/kubernetes/kubernetes/pull/16751/files#diff-97a1ffdd78274d54ba9dc28656ceca2dR149 ?